### DISCUSSION TITLE
RECESSIVE search is getting reset to HOMOZYGOUS_RECESSIVE

### DISCUSSION THREAD

**Discussion by NLSVTN** [broadinstitute/seqr#2339] [Q&A]
If we select **RECESSIVE** search, then run it, change a single parameter on the form (say, frequency) and resubmit the form, **RECESSIVE** in the form changes to **HOMOZYGOUS_RECESSIVE**. I doubt it should happen because it would confuse the user who would think that the search was **RECESSIVE** but then sees **HOMOZYGOUS_RECESSIVE**. The issue happens because of the following line:

https://github.com/broadinstitute/seqr/blob/master/seqr/utils/elasticsearch/es_search.py#L384

And the search is later on returned and the form is updated with the new inheritance filter value. Shouldn't we copy **inheritance_filter** instead to prevent such behavior?

**Comment by hanars** [comment]
We are not able to reproduce this bug in our seqr instance. Following the steps you specify, the inheritance never changes unless you explicitly change it.  The file you link is unrelated to the client-side form rendering so it is unlikely that making a change there would have any impact, but you are welcome to try changing it yourselves to see

**Reply by NLSVTN** [reply]
It’s not reproducible because redis stores cache data most probably. One way to try is to clean redis and try one more time. It happens only with recessive. 

Yes, it is backend file. What happens is the following: there is ‘search’ object that you pass down and modify accidentally because no copy is done. Then, this ‘search’ object is passed back to the front end and redux does the corresponding update.

**Reply by hanars** [reply]
Unfortunately,  I still can not replicate this issue on our end. When I do the work flow you describe, (doing a recessive search, waiting for results, changing one non-inheritance parameter, and hitting search again), I still see "recessive" selected as the inheritance mode and I get back all recessive search results. As such I can not fix this bug for you. I don't think the way the client determines what inheritance to show works the way you think it does. Please feel free to make whatever changes are appropriate for you on your own branch.

**Reply by NLSVTN** [reply]
I see, could be some of our changes too. I will double check the code. The bug is fixed on our side, just thought to let you know. If you don’t see it, it’s probably fine.

**Comment by NLSVTN** [comment]
In **get_es_variants()** function 'results_model' is passed by reference:
 
https://github.com/broadinstitute/seqr/blob/master/seqr/views/apis/variant_search_api.py#L59
 
There we get **search** dict:
 
https://github.com/broadinstitute/seqr/blob/02e49ccd79e586a5e270ab60db724480b535068b/seqr/utils/elasticsearch/utils.py#L100
 
Then, **inheritance** part of the **search** dict is passed down to **filter_by_annotation_and_genotype() -> _filter_by_genotype() -> _get_family_sample_query()** and up to the line that I posted above:
 
https://github.com/broadinstitute/seqr/blob/master/seqr/utils/elasticsearch/es_search.py#L384 
 
So, the result is that if any part of the **search** is modified by **get_es_variants()** function (and it does so for us), then **search** object will be returned in its modified form:
 
https://github.com/broadinstitute/seqr/blob/master/seqr/views/apis/variant_search_api.py#L76
 
And here the returned **search** object is updated in **Redux** state:
 
https://github.com/broadinstitute/seqr/blob/master/ui/redux/rootReducer.js#L219
 
After that, since **filter** inside **search.inheritance** got modified by the backend the following code switches **recessive** for us:
 
https://github.com/broadinstitute/seqr/blob/master/ui/pages/Search/components/VariantSearchFormContent.jsx#L76
 
I don't understand the theory precisely behind when such modification of filter happens on backend, but since it does for us, it may do such a thing on some projects and with some form parameters in the future for you too. I think it’s a dangerous thing to do, to bind backend to frontend. It seems to me better if the form is managing itself and backend is not affecting it (if its not a bug).
 
We fixed it by adding a single line of code:
 
```
    If inheritance_mode:
           # copying inheritance_filter, otherwise search object will be modified which causes
           # bugs in variant search form (e.g. RECESSIVE -> HOMOZYGOUS_RECESSIVE switch)
    -----> inheritance_filter=dict(inheritance_filter)
           inheritance_filter.update(INHERITANCE_FILTERS[inheritance_mode])
```
 
Perhaps it's not the best fix and we should instead do a deep copy of the whole passed 'search' object but for now decided to go with that.

**Reply by hanars** [reply]
I'm glad this fixed it for you. Feel free to put up pull requests for my review and I can merge your fixes to our master branch and share them with other users.  And in the future, when you encounter bugs like this where you have already fixed the problem locally, you should really feel free to put up pull requests for my review without even submitting an issue. The turnaround time will be much faster for me to review a fix than for me to need to independently research and fix a small bug. Remeber that our team is really small so I can't always provide timely support for these sorts of things

I also really still can't replicate this bug, I think our state works differently than it does for you - we use the search provided by the user and not the one returned from the server in most cases. The search returned from the server is used only when a user accesses a search results link created in a different browser/ by a different user where they don't have a copy of the search cached in their browser. Maybe you have local storage disabled or are using a different browser with different default behavior or something like that which is why you are seeing the bug and I am not

**Reply by NLSVTN** [reply]
Oh, ok got it, will do. 

Hmm, interesting, need to think about it.

Thank you!
